File: react-frontend/src/api/endpoints/super-admin.ts
/**
 * Super Admin API endpoints
 */

import { apiClient } from '../client';

const BASE_PATH = '/api/v1/super-admin';

// Types
export interface SuperAdmin {
  id: number;
  email: string;
  name: string;
  is_active: boolean;
  mfa_enabled: boolean;
  last_login_at: string | null;
  last_login_ip: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface SuperAdminLoginRequest {
  email: string;
  password: string;
  mfa_token?: string;
}

export interface SuperAdminLoginResponse {
  access_token: string;
  token_type: string;
  requires_mfa: boolean;
  super_admin: SuperAdmin;
}

export interface MFASetupResponse {
  secret: string;
  provisioning_uri: string;
  backup_codes: string[];
}

export interface ImpersonationRequest {
  tenant_id: number;
  target_user_id?: number;
  read_only?: boolean;
}

export interface ImpersonationResponse {
  access_token: string;
  token_type: string;
  tenant_id: number;
  tenant_name: string;
  target_user_id: number;
  target_username: string;
  read_only: boolean;
  expires_at: string;
}

export interface TenantSummary {
  id: number;
  name: string;
  display_name: string;
  subscription_status: string;
  plan_name: string | null;
  users_count: number;
  studies_count: number;
  created_at: string;
  trial_ends_at: string | null;
  is_deleted: boolean;
}

export interface TenantListResponse {
  items: TenantSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface DashboardStats {
  total_tenants: number;
  active_tenants: number;
  trialing_tenants: number;
  past_due_tenants: number;
  canceled_tenants: number;
  total_users: number;
  total_studies: number;
  revenue_mrr: number;
}

// Billing Types
export interface TenantBillingDetails {
  id: number;
  name: string;
  display_name: string;
  created_at: string;
  stripe_customer_id: string | null;
  stripe_subscription_id: string | null;
  stripe_dashboard_url: string | null;
  subscription_status: string;
  plan_name: string | null;
  plan_id: number | null;
  current_period_start: string | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
  trial_ends_at: string | null;
  grace_period_ends_at: string | null;
  users_count: number;
  users_limit: number;
  studies_count: number;
  studies_limit: number;
}

export interface InvoiceSummary {
  id: string;
  number: string | null;
  status: string;
  amount_due: number;
  amount_paid: number;
  currency: string;
  created: string;
  due_date: string | null;
  paid_at: string | null;
  invoice_pdf: string | null;
  hosted_invoice_url: string | null;
}

export interface TenantBillingResponse {
  tenant: TenantBillingDetails;
  invoices: InvoiceSummary[];
}

export interface RevenueByPlan {
  plan_name: string;
  plan_id: number;
  tenant_count: number;
  mrr: number;
}

export interface RevenueStats {
  total_mrr: number;
  by_plan: RevenueByPlan[];
  trial_count: number;
  trial_conversion_rate: number | null;
  churned_last_30_days: number;
}

// Super admin token storage
// Using sessionStorage for enhanced security - token doesn't persist across browser sessions
const SUPER_ADMIN_TOKEN_KEY = 'super_admin_token';

export const setSuperAdminToken = (token: string) => {
  // Use sessionStorage for super admin tokens (more secure than localStorage)
  sessionStorage.setItem(SUPER_ADMIN_TOKEN_KEY, token);
};

export const getSuperAdminToken = (): string | null => {
  return sessionStorage.getItem(SUPER_ADMIN_TOKEN_KEY);
};

export const clearSuperAdminToken = () => {
  sessionStorage.removeItem(SUPER_ADMIN_TOKEN_KEY);
};

// Helper to get auth header for super admin
const getSuperAdminAuthHeader = () => {
  const token = getSuperAdminToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
};

// API Functions

/**
 * Super admin login
 */
export const superAdminLogin = async (data: SuperAdminLoginRequest): Promise<SuperAdminLoginResponse> => {
  const response = await apiClient.post<SuperAdminLoginResponse>(`${BASE_PATH}/login`, data);
  if (response.data.access_token && !response.data.requires_mfa) {
    setSuperAdminToken(response.data.access_token);
  }
  return response.data;
};

/**
 * Get current super admin profile
 */
export const getSuperAdminMe = async (): Promise<SuperAdmin> => {
  const response = await apiClient.get<SuperAdmin>(`${BASE_PATH}/me`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Setup MFA for super admin
 */
export const setupMFA = async (): Promise<MFASetupResponse> => {
  const response = await apiClient.post<MFASetupResponse>(`${BASE_PATH}/mfa/setup`, {}, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Verify MFA setup
 */
export const verifyMFA = async (token: string): Promise<{ message: string }> => {
  const response = await apiClient.post(`${BASE_PATH}/mfa/verify`, { token }, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get dashboard statistics
 */
export const getDashboardStats = async (): Promise<DashboardStats> => {
  const response = await apiClient.get<DashboardStats>(`${BASE_PATH}/dashboard/stats`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get list of tenants
 */
export const getTenants = async (params?: {
  page?: number;
  page_size?: number;
  status_filter?: string;
  search?: string;
}): Promise<TenantListResponse> => {
  const response = await apiClient.get<TenantListResponse>(`${BASE_PATH}/tenants`, {
    params,
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get a specific tenant
 */
export const getTenant = async (tenantId: number): Promise<TenantSummary> => {
  const response = await apiClient.get<TenantSummary>(`${BASE_PATH}/tenants/${tenantId}`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Start impersonation session
 */
export const startImpersonation = async (data: ImpersonationRequest): Promise<ImpersonationResponse> => {
  const response = await apiClient.post<ImpersonationResponse>(`${BASE_PATH}/impersonate`, data, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * End impersonation session (for audit logging)
 */
export const endImpersonation = async (tenantId?: number): Promise<{ message: string }> => {
  const response = await apiClient.post(`${BASE_PATH}/impersonate/end`, 
    tenantId ? { tenant_id: tenantId } : {},
    { headers: getSuperAdminAuthHeader() }
  );
  return response.data;
};

/**
 * Logout super admin
 */
export const superAdminLogout = () => {
  clearSuperAdminToken();
};

// Helper functions

/**
 * Format subscription status for display
 */
export const getStatusBadge = (status: string): { label: string; variant: 'default' | 'secondary' | 'destructive' | 'outline' } => {
  switch (status) {
    case 'active':
      return { label: 'Active', variant: 'default' };
    case 'trialing':
      return { label: 'Trial', variant: 'secondary' };
    case 'past_due':
      return { label: 'Past Due', variant: 'destructive' };
    case 'canceled':
      return { label: 'Canceled', variant: 'outline' };
    case 'unpaid':
      return { label: 'Unpaid', variant: 'destructive' };
    default:
      return { label: status, variant: 'secondary' };
  }
};

/**
 * Format MRR for display
 */
export const formatMRR = (cents: number): string => {
  return `$${(cents / 100).toLocaleString('en-US', { minimumFractionDigits: 0, maximumFractionDigits: 0 })}`;
};

/**
 * Format currency amount in cents for display
 */
export const formatCurrency = (cents: number, currency: string = 'usd'): string => {
  const amount = cents / 100;
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency.toUpperCase(),
    minimumFractionDigits: 2,
  }).format(amount);
};

/**
 * Get invoice status badge styling
 */
export const getInvoiceStatusBadge = (status: string): { label: string; variant: 'default' | 'secondary' | 'destructive' | 'outline' } => {
  switch (status) {
    case 'paid':
      return { label: 'Paid', variant: 'default' };
    case 'open':
      return { label: 'Open', variant: 'secondary' };
    case 'draft':
      return { label: 'Draft', variant: 'outline' };
    case 'void':
      return { label: 'Void', variant: 'outline' };
    case 'uncollectible':
      return { label: 'Uncollectible', variant: 'destructive' };
    default:
      return { label: status, variant: 'secondary' };
  }
};

// Billing API Functions

/**
 * Get detailed billing information for a tenant
 */
export const getTenantBilling = async (tenantId: number): Promise<TenantBillingResponse> => {
  const response = await apiClient.get<TenantBillingResponse>(`${BASE_PATH}/tenants/${tenantId}/billing`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get revenue breakdown by plan
 */
export const getRevenueBreakdown = async (): Promise<RevenueStats> => {
  const response = await apiClient.get<RevenueStats>(`${BASE_PATH}/revenue/breakdown`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

// Feature Request Types
export interface FeatureRequestWithUser {
  id: number;
  tenant_id: number;
  user_id: number;
  category: 'bug' | 'feature';
  title: string;
  description: string;
  status: 'pending' | 'working' | 'done';
  admin_response?: string;
  responded_at?: string;
  created_at: string;
  updated_at: string;
  user_name?: string;
  user_email?: string;
  tenant_name?: string;
}

export interface FeatureRequestListResponse {
  items: FeatureRequestWithUser[];
  total: number;
  skip: number;
  limit: number;
}

export interface FeatureRequestStats {
  pending: number;
  working: number;
  done: number;
  total: number;
}

export interface FeatureRequestUpdate {
  status?: 'pending' | 'working' | 'done';
  admin_response?: string;
}

// Feature Request API Functions

/**
 * Get all feature requests across all tenants
 */
export const getFeatureRequests = async (params?: {
  skip?: number;
  limit?: number;
  status_filter?: string;
  category_filter?: string;
}): Promise<FeatureRequestListResponse> => {
  const response = await apiClient.get<FeatureRequestListResponse>(`${BASE_PATH}/feature-requests`, {
    params,
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get feature request statistics
 */
export const getFeatureRequestStats = async (): Promise<FeatureRequestStats> => {
  const response = await apiClient.get<FeatureRequestStats>(`${BASE_PATH}/feature-requests/stats`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Get a specific feature request
 */
export const getFeatureRequest = async (requestId: number): Promise<FeatureRequestWithUser> => {
  const response = await apiClient.get<FeatureRequestWithUser>(`${BASE_PATH}/feature-requests/${requestId}`, {
    headers: getSuperAdminAuthHeader(),
  });
  return response.data;
};

/**
 * Update a feature request (status and/or admin response)
 */
export const updateFeatureRequest = async (
  requestId: number,
  data: FeatureRequestUpdate
): Promise<FeatureRequestWithUser> => {
  const response = await apiClient.put<FeatureRequestWithUser>(
    `${BASE_PATH}/feature-requests/${requestId}`,
    data,
    { headers: getSuperAdminAuthHeader() }
  );
  return response.data;
};

/**
 * Get feature request status badge styling
 */
export const getFeatureRequestStatusBadge = (status: string): { label: string; variant: 'default' | 'secondary' | 'destructive' | 'outline'; color: string } => {
  switch (status) {
    case 'pending':
      return { label: 'Pending', variant: 'secondary', color: 'amber' };
    case 'working':
      return { label: 'Working', variant: 'default', color: 'blue' };
    case 'done':
      return { label: 'Done', variant: 'outline', color: 'green' };
    default:
      return { label: status, variant: 'secondary', color: 'gray' };
  }
};

/**
 * Get feature request category badge styling
 */
export const getFeatureRequestCategoryBadge = (category: string): { label: string; variant: 'default' | 'secondary' | 'destructive' | 'outline'; icon: string } => {
  switch (category) {
    case 'bug':
      return { label: 'Bug', variant: 'destructive', icon: 'bug' };
    case 'feature':
      return { label: 'Feature', variant: 'secondary', icon: 'lightbulb' };
    default:
      return { label: category, variant: 'secondary', icon: 'help' };
  }
};

File: backend/app/schemas/super_admin.py
"""Pydantic schemas for Super Admin."""

from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, EmailStr, Field, ConfigDict


# =============================================================================
# Authentication Schemas
# =============================================================================

class SuperAdminLogin(BaseModel):
    """Login request for super admin."""
    email: str = Field(..., min_length=1, max_length=255, description="Super admin email address")
    password: str = Field(..., min_length=8)
    mfa_token: Optional[str] = Field(None, description="MFA token if MFA is enabled")


class SuperAdminLoginResponse(BaseModel):
    """Login response with token."""
    access_token: str
    token_type: str = "bearer"
    requires_mfa: bool = False
    super_admin: "SuperAdminRead"


class MFASetupResponse(BaseModel):
    """Response when setting up MFA."""
    secret: str
    provisioning_uri: str
    backup_codes: List[str]


class MFAVerifyRequest(BaseModel):
    """Request to verify MFA setup or login."""
    token: str = Field(..., min_length=6, max_length=6)


# =============================================================================
# Impersonation Schemas
# =============================================================================

class ImpersonationRequest(BaseModel):
    """Request to impersonate a tenant user."""
    tenant_id: int
    target_user_id: Optional[int] = Field(
        None, 
        description="User to impersonate (defaults to tenant admin)"
    )
    read_only: bool = Field(
        True, 
        description="If true, mutations are blocked"
    )


class ImpersonationResponse(BaseModel):
    """Response with impersonation token."""
    access_token: str
    token_type: str = "bearer"
    tenant_id: int
    tenant_name: str
    target_user_id: int
    target_username: str
    read_only: bool
    expires_at: datetime


# =============================================================================
# Super Admin CRUD Schemas
# =============================================================================

class SuperAdminCreate(BaseModel):
    """Schema for creating a super admin."""
    email: EmailStr
    password: str = Field(..., min_length=8)
    name: str = Field(..., min_length=2, max_length=255)


class SuperAdminUpdate(BaseModel):
    """Schema for updating a super admin."""
    name: Optional[str] = Field(None, min_length=2, max_length=255)
    is_active: Optional[bool] = None


class SuperAdminRead(BaseModel):
    """Schema for reading a super admin."""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    email: str
    name: str
    is_active: bool
    mfa_enabled: bool
    last_login_at: Optional[datetime] = None
    last_login_ip: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None


class SuperAdminList(BaseModel):
    """List of super admins."""
    items: List[SuperAdminRead]
    total: int


# =============================================================================
# Tenant Management Schemas (for Super Admin Dashboard)
# =============================================================================

class TenantSummary(BaseModel):
    """Summary of a tenant for the super admin dashboard."""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    name: str
    display_name: str
    subscription_status: str
    plan_name: Optional[str] = None
    users_count: int
    studies_count: int
    created_at: datetime
    trial_ends_at: Optional[datetime] = None
    is_deleted: bool


class TenantListResponse(BaseModel):
    """List of tenants with pagination."""
    items: List[TenantSummary]
    total: int
    page: int
    page_size: int


class DashboardStats(BaseModel):
    """Dashboard statistics for super admin."""
    total_tenants: int
    active_tenants: int
    trialing_tenants: int
    past_due_tenants: int
    canceled_tenants: int
    total_users: int
    total_studies: int
    revenue_mrr: int  # Monthly recurring revenue in cents


# =============================================================================
# Billing Management Schemas (for Super Admin)
# =============================================================================

class TenantBillingDetails(BaseModel):
    """Extended billing details for a tenant."""
    model_config = ConfigDict(from_attributes=True)

    # Basic tenant info
    id: int
    name: str
    display_name: str
    created_at: datetime

    # Stripe identifiers
    stripe_customer_id: Optional[str] = None
    stripe_subscription_id: Optional[str] = None
    stripe_dashboard_url: Optional[str] = None

    # Subscription details
    subscription_status: str
    plan_name: Optional[str] = None
    plan_id: Optional[int] = None

    # Billing period (from Stripe)
    current_period_start: Optional[datetime] = None
    current_period_end: Optional[datetime] = None
    cancel_at_period_end: bool = False

    # Trial/Grace
    trial_ends_at: Optional[datetime] = None
    grace_period_ends_at: Optional[datetime] = None

    # Usage
    users_count: int
    users_limit: int  # -1 = unlimited
    studies_count: int
    studies_limit: int  # -1 = unlimited


class InvoiceSummary(BaseModel):
    """Summary of a Stripe invoice."""
    id: str
    number: Optional[str] = None
    status: str  # draft, open, paid, void, uncollectible
    amount_due: int  # in cents
    amount_paid: int  # in cents
    currency: str
    created: datetime
    due_date: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    invoice_pdf: Optional[str] = None
    hosted_invoice_url: Optional[str] = None


class TenantBillingResponse(BaseModel):
    """Full billing response for a tenant."""
    tenant: TenantBillingDetails
    invoices: List[InvoiceSummary]


class RevenueByPlan(BaseModel):
    """Revenue breakdown by plan."""
    plan_name: str
    plan_id: int
    tenant_count: int
    mrr: int  # Monthly recurring revenue in cents


class RevenueStats(BaseModel):
    """Extended revenue statistics for platform."""
    total_mrr: int
    by_plan: List[RevenueByPlan]
    trial_count: int
    trial_conversion_rate: Optional[float] = None  # percentage (0-100)
    churned_last_30_days: int


# Forward references
SuperAdminLoginResponse.model_rebuild()
